workflows/crisprseq.nf

```diff
@@ -71,16 +97,329 @@ workflow CRISPRSEQ {
+    //
+    CAT_FASTQ (
+        ch_fastq.multiple
+    )
+    .reads
+    .groupTuple(by: [0])
+    .mix(ch_fastq.single)
+    // Separate samples by paired-end or single-end
+    .branch {
+        meta, fastq ->
+            single: fastq.size() == 1
+                return [ meta, fastq ]
+            paired: fastq.size() > 1
+                return [ meta, fastq ]
+    }
+    .set { ch_cat_fastq }
+    // comment version as there is an error (cat version is "cat: cat:")
+    //ch_versions = ch_versions.mix(CAT_FASTQ.out.versions.first().ifEmpty(null))
+
+    //
+    // MODULE: Merge paired-end reads
+    //
+    PEAR (
+        ch_cat_fastq.paired
+    )
+    .assembled
+    .mix( ch_cat_fastq.single )
+    .set { ch_pear_fastq }
+    ch_versions = ch_versions.mix(PEAR.out.versions)
+
     //
     // MODULE: Run FastQC
     //
     FASTQC (
-        INPUT_CHECK.out.reads
+        ch_pear_fastq
     )
     ch_versions = ch_versions.mix(FASTQC.out.versions.first())
 
+    FIND_ADAPTERS (
+        FASTQC.out.zip
+    )
+    .adapters
+    .join(ch_pear_fastq)
+    .groupTuple(by: [0])
+    // Separate samples by containing overrepresented sequences or not
+    .branch {
+        meta, adapter_seqs, reads ->
+            no_adapters: adapter_seqs[0].size() == 0
+                return [ meta, reads[0] ]
+            adapters   : adapter_seqs[0].size() > 0
+                return [ meta, adapter_seqs[0], reads[0] ]
+    }
+    .set { ch_adapter_seqs }
+
+
+    //
+    // MODULE: Trim adapter sequences
+    //
+    CUTADAPT (
+        ch_adapter_seqs.adapters
+    )
+
+    ch_adapter_seqs.no_adapters
+    .mix(CUTADAPT.out.reads)
+    .groupTuple(by: [0])
+    .map {
+        meta, fastq ->
+                return [ meta, fastq.flatten() ]
+    }
+    .set{ ch_trimmed }
+
+    //
+    // MODULE: Mask (convert to Ns) bases with quality lower than 20 and remove sequences shorter than 80
+    //
+    SEQTK_SEQ (
+        ch_trimmed
+    )
+
+
+    MERGING_SUMMARY {
+        ch_cat_fastq.paired
+            .mix(ch_cat_fastq.single)
+            .join(PEAR.out.assembled, remainder: true)
+            .join(SEQTK_SEQ.out.fastx)
+            .join(CUTADAPT.out.log)
+            .map { meta, reads, assembled, masked, trimmed ->
+                if (assembled == null) {
+                    dummy = file('null')
+                    return [ meta, reads, dummy, masked, trimmed ]
+                } else {
+                    return [ meta, reads, assembled, masked, trimmed ]
+                }
+            }
+    }
+
+
+    /*
+    Remove this step by now as I don't have test data
+    //
+    // MODULE: Extract UMI sequences
+    //
+    EXTRACT_UMIS (
+        SEQTK_SEQ.out.fastx
+    )
+
+    To implement for UMIs
+
+    vsearch
+    get_ubs
+    top_read
+    polishing: minimap2, racon
+    consensus
+    join_reads
+    fa2fq
+
+    */
+
+    // Dummy final UMI reads fastq
+    DUMMY_FINAL_UMI {
+        SEQTK_SEQ.out.fastx
+    }
```

Comment:
Not sure I followed the whole UMI logic there

Reply:
I improved the comment. This will be implemented in the future, but I leave the structure annotated here.